_posts/drafts/exploring-go-objects-part2.md

```diff
@@ -1,76 +1,411 @@
+
+* Express clearly what is the protocol it needs
+* Be sure that when it starts interacting with an implementation no function is missing
+
+The object that implements the service needs to:
+
+* Be able to safely express that it has the required functions to satisfy the protocol
+* Be able to satisfy a protocol even without knowing it explicitly
+
+There is no need for two objects to interact to achieve a common goal
+to have a specific type, all that matters is if the protocol between
+them matches.
+
+This is where Go interfaces comes in.
+It provides us with a compile time safe way to express protocols eliminating
+all the boilerplate of initializing structs with the proper functions.
+It will initialize the structs for us, and even optimize to
+initialize the struct just once, what is called in Go the **iface**,
+which is comparable to C++ **vtable**.
+
+It will also allow the code to be more decoupled since you don't have
+to known the package where the interface is defined to implement it.
+This is fundamental to how Go allows more flexibility with the same
+compile time safety as languages like Java/C++.
+
+Lets revisit the previous file protocol with interfaces:
 
 ```go
-type SomeProtocol interface {
-    Do(a string) error
-    DoOther(a int) (int, error)
+package main
+
+type Reader interface {
+    Read(data []byte) (int, error)
+}
+
+type Closer interface {
+    Close() error
+}
+
+type ReadCloser interface {
+    Reader
+    Closer
+}
+
+type Opener func() (ReadCloser, error)
+
+type File struct {}
+
+func (f *File) Read(data []byte) (int, error) {
+        return 0, nil
 }
 
-func useSomeProtocol(s SomeProtocol) {
-        s.Do("hi")
-        s.DoOther(5)
+func (f *File) Close() error {
+        return nil
+}
+
+func useFileProtocol(open Opener) {
+        f, _ := open()
+        data := make([]byte, 50)
+        f.Read(data)
+        f.Close()
 }
 
 func main() {
-        useSomeProtocol(SomeProtocol{
-            Do: func(a string) error {
-                    return nil
-            },
-            DoOther: func(a int) (int, error) {
-                    return 0, nil
-            },
+        useFileProtocol(func() (ReadCloser, error) {
+                return &File{}, nil
         })
 }
 ```
 
-The negative point of this is that there is no easy way to verify at
-compile time that all the functions of the interface instance
-have been initialized with a proper function.
+One key difference is that with interfaces this code is now
+safe. The **useFileProtocol** do not have to worry with calling
+functions that are actually nil, the Go compiler will do the work
+of creating a struct that holds a pointer with a **iface** descriptor
+that has all the functions required to satisfy the protocol.
+It does this per match of type <-> interface, as it is used (initializes
+on first usage). 
+
+You still can cause a segmentation fault if you do something like this:
+
+```go
+func() (ReadCloser, error) {
```

Comment:
the intention isn't clear in a top-level declaration (and does not works also), better to make the intention clear as in:

```go
useFileProtocol(func() (ReadCloser, error) {
    var a ReadCloser
    return a, nil
})
```
what do you think?

Reply:
nicer indeed